auth/authzn.proto

```diff
@@ -0,0 +1,179 @@
+//
+// Copyright 2020 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+syntax = "proto3";
+
+import "github.com/openconfig/gnoi/types/types.proto";
+
+// This package is a work in progress. It may change in the future. :-)
+package gnoi.authzn;
+
+option (types.gnoi_version) = "0.1.0";
+
+// Service Authzn supplies the target (i.e., network device, server end of the
+// gRPC connection) with a definition of an authentication and authorization
+// policy that should be used for the respective (authn/authz) of the relevant
+// service. 
+//
+// It is expected that when authorization and authentication details are
+// supplied via the gNOI Authzn service, this config cannot be manipulated
+// through other interfaces, such that a restriction on the gNOI.authzn RPC
+// itself restricts the set of users that are able to manipulate the auth[zn]
+// configuration of the end device.
+//
+// In order to ensure that the auth[zn] configuration supplied in the service
+// cannot be subverted by changing the configuration of the device via gNMI,
+// the following leaves with the OpenConfig system model should be ignored:
+// 
+//  - /system/aaa/authentication/config/authentication-method
+//  - /system/aaa/authorization/config/authorization-method
+//  - /system/aaa/authentication/admin-user/...
+//  - /system/aaa/authentication/users/...
+//
+// The applied state (i.e., the 'state' version of the paths above) should
+// reflect a new authentication-method and authorization-method of GNOI_AUTHZN
+// which is to be added as a new identity with the base of
+// openconfig-aaa-types:AAA_METHOD_TYPE (http://ops.openconfig.net/branches/models/master/docs/openconfig-system.html#ident-aaa_method_type)
+//
+// The motivation for this service is provided in the accompanying
+// authn-authz-intro.md document.
+service Authzn {
+    // Set allows a user to specify an authentication and authorization
+    // policy to the network element. The policy, if valid, replaces the
+    // existing policy.
+    //
+    // Canonical error codes are to be used as responses. See
+    // https://grpc.io/docs/guides/error/ for more details.
+    rpc Set(SetRequest) returns (SetResponse) {}
+}
+
+message SetRequest {
```

Comment:
How about removing a policy? Is the plan that the complete set of policies has to be provided? For all types of access?

Reply:
This would be my suggestion, yes - such that we don't have to add some form of 'Get' API to retrieve the current AAA proto, and rather just declaratively update it.